### PR TITLE
Caching InputLine calculation. Caching InputArea::height.

### PR DESCRIPTION
**Pull request by trevarj**
First attempt at solving some of the issues on #208

### CODE REVIEW

**Comment by osa1** [comment]
Thanks for doing this!

I did a quick review. Questions/comments:

- Do we really need all these fields in `LineDataCache`? For example, as far as I understand it doesn't make sense to re-use `last_whitespace_idx` and `prev_char_is_whitespace` from a previous call to `calculate_height`, so those states should be local to `calcualte_height`.

- Related to the point above: if you document these fields for what they are/ how they are used/ why they are needed the it may become clear which ones are needed in this type and which ones could be moved to the function `calculate_height` as a local variable.

- Methods of `LineDataCache` don't need to be `pub`.

- If I add a character to the end of the buffer we still run the whole computation from scratch, right? I think we could have a special method for inserting at the end that runs much faster. 

Btw, I realized that my remark about quadratic behavior on paste in #208 is incorrect: we call `TUI::keypressed` for every char in the pasted string, but `keypressed` doesn't draw after handling the key. So we call `draw` just once after handling all the keypresses. Sorry for the false alarm!

**Comment by trevarj** [comment]
> Do we really need all these fields in LineDataCache? For example, as far as I understand it doesn't make sense to re-use last_whitespace_idx and prev_char_is_whitespace from a previous call to calculate_height, so those states should be local to calcualte_height.

I do believe we need those variables in order to save the state of the calculation, due to the "fast track" scenario, where we insert into the end of the buffer. It allows for resuming the calculation at the point where we left off.

> Related to the point above: if you document these fields for what they are/ how they are used/ why they are needed the it may become clear which ones are needed in this type and which ones could be moved to the function calculate_height as a local variable.

I can do this.

> Methods of LineDataCache don't need to be pub.

Done.

> If I add a character to the end of the buffer we still run the whole computation from scratch, right? I think we could have a special method for inserting at the end that runs much faster.

Same as explanation above, but here is the change to insert()
```
    /// Interface for Vec::insert()
    pub(crate) fn insert(&mut self, idx: usize, element: char) {
        self.buffer.insert(idx, element);
        if idx == self.buffer.len() - 1 {
            self.line_data.calculate_height(&self.buffer, idx);
        } else {
            self.line_data.set_dirty();
        }
    }
```
Note that when the idx is the end of the buffer, we resume the calculation from where it ended before. The call to `calculate_height` with an `offset` of `idx` will look at the state and decide if an index needs to be added to `ws_indices`. Maybe the `offset` idea isn't so great, and it could just be a boolean like `append` or something.

**Comment by trevarj** [comment]
Rebased with master now build fails

File: libtiny_tui/src/input_area/input_line.rs [hunk 1]
```diff
@@ -2,30 +2,175 @@ use crate::{config::Colors, termbox, utils};
 use std::{cmp::min, ops::RangeBounds, vec::Drain};
 use termbox_simple::Termbox;
 
+/// Cache that stores the state of InputLine's height calculation.
+/// `line_count` is used as the dirty bit to invalidate the cache.
+#[derive(Clone, Debug)]
+struct LineDataCache {
+    /// Indices to split on when we draw, not always whitespaces
+    ws_indices: Vec<i32>,
```
Comment (osa1) [inline libtiny_tui/src/input_area/input_line.rs:10]: Ahh, so you also fix #202 with this, right? Let's rename this to `split_indices`.

Comment (trevarj) [reply libtiny_tui/src/input_area/input_line.rs:10]: No, it doesn't fix that #202, sadly. That issue existed before any of the input changes. I've been thinking about how to fix that problem though.

I will rename.

Comment (osa1) [reply libtiny_tui/src/input_area/input_line.rs:10]: Ah, I'm completely confused. This is about the input field of course it doesn't fix #202.

**Comment by osa1** [comment]
Thanks @trevarj , this looks good to me. I tested this a little bit and I can see that caching works, and we update the state after the insertion point (instead of creating it from scratch) on key presses and also when pasting.

We could potentially add one more special case for deleting from the end (in `InputLine::remove`), but this is fine for now.

**Pull request merged**